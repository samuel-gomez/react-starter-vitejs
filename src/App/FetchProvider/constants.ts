export const REFETCH_ON_WINDOWS_FOCUS = false;
export const RETRY_REQUEST = 2;
export const CACHE_TIME = 1000 * 60 * 30; // 30 min de cache
