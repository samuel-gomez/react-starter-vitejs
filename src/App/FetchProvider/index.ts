export { default, FetchContext } from './FetchProvider';
export type { FetchContextType, TFetchCustom } from './FetchProvider';
