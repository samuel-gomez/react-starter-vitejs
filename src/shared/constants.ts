import packageJson from '../../package.json';

/** ***********************************************************************
 * FORM CONSTANTS
 ************************************************************************ */
export const MSG_REQUIRED = 'Le champ est obligatoire';

export const MSG_LENGTH_3 = 'Le champ doit faire au moins 3 caractères';
export const MSG_FORMAT = 'Format invalide';
export const MSG_ERROR_FORMAT = 'Format de fichier incorrect';

export const DEFAULT_OPTION_LABEL = '- Sélectionner -';
export const DEFAULT_OPTION = {
  value: '',
  label: DEFAULT_OPTION_LABEL,
};

// patterns
export const patternEmail = /^[\w-+]+@([\w-+]+\.)+[\w-]{2,}$/;
export const patternPhoneNumber = /^[+0][\d ]*$/;
export const patternZipCode = /^\d{5}$/;
export const patternSiret = /^\d{14}$/;
export const patternSiren = /^\d{9}$/;
export const patternSiretSiren = /^\d{9,14}$/;

export const phoneNumberRule = {
  required: true,
  minLength: 4,
  maxLength: 20,
  pattern: patternPhoneNumber,
};

export const emailRule = {
  required: true,
  minLength: 6,
  maxLength: 60,
  pattern: patternEmail,
};

/** ***********************************************************************
 * API STATUS CONSTANTS
 ************************************************************************ */
export const STATUS_HTTP = {
  SUCCESS: 200,
  CREATED: 201,
  BAD_REQUEST: 400,
  UNAUTHORIZE: 401,
  FORBIDDEN: 403,
  NOTFOUND: 404,
  SERVER_ERROR: 500,
};

export const STATUS_API = {
  SUCCESS: STATUS_HTTP.SUCCESS,
  CREATED: STATUS_HTTP.CREATED,
  WARNING: 40,
  ERROR: 50,
};

/** ***********************************************************************
 * API URL
 ************************************************************************ */
export const API_URL = {
  BASE: 'base',
  GITHUB: 'github',
};

export const STATUS_HTTP_MESSAGES = {
  [STATUS_HTTP.SUCCESS]: 'Succès: La requête a réussi',
  [STATUS_HTTP.CREATED]: 'Succès: Création réussie',
  [STATUS_HTTP.BAD_REQUEST]: 'Erreur: La syntaxe de la requête est erronée',
  [STATUS_HTTP.UNAUTHORIZE]: 'Erreur: Une authentification est nécessaire pour accéder à la ressource',
  [STATUS_HTTP.FORBIDDEN]: 'Erreur: Accès non autorisé',
  [STATUS_HTTP.NOTFOUND]: 'Erreur: Elément non trouvé',
  [STATUS_HTTP.SERVER_ERROR]: 'Erreur: Problème technique ! Contacter votre support',
};

/** ***********************************************************************
 * USER CONSTANTS
 ************************************************************************ */

export const ADMIN = 'Admin';
export const READER = 'Reader';
export const WRITER = 'Writer';
export const USER = 'User';
export const ALL = '';

export const PROFILS = [ADMIN, READER, WRITER, USER, ALL];

/** ***********************************************************************
 * COMMONS CONSTANTS
 ************************************************************************ */

export const MODIFIER_CLASS = {
  success: 'success',
  disabled: 'disabled',
};

/** ***********************************************************************
 * VERSION CONSTANT
 ************************************************************************ */

const VERSION_TOOLKIT = 'master';

/** ***********************************************************************
 * DEMO CONSTANTS
 ************************************************************************ */

export const STORYBOOK = 'https://axafrance.github.io/react-toolkit/latest/storybook/?path=/story/';
export const DESIGN_SYSTEM = 'https://axafrance.github.io/design-system/';
export const GITHUB = `https://github.com/AxaGuilDEv/react-toolkit/tree/v${packageJson.dependencies['@axa-fr/react-toolkit-all']}/packages/`;
export const GITHUB_API = `AxaGuilDEv/react-toolkit/${VERSION_TOOLKIT}/packages/`;
