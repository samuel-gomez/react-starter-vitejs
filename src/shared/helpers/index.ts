export { formatDate, isValidDate, setDate } from './formatDate';
export { default as isEmptyOrNull, isEmptyArray, isEmptyObject } from './isEmptyOrNull';
export { default as mergeObj } from './mergeObj';
export { default as setAnomalyEmptyItems } from './setAnomalyEmptyItems';
export { default as setConfirmClassModifier } from './setConfirmClassModifier';
export { default as emptyFunction } from './emptyFunction';
export { default as setRequiredFieldRule, setRequiredMessage } from './setRequiredFieldRule';
