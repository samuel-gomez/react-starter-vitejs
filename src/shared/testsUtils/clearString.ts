const cleanString = (str?: string) => str?.replace(/\s/g, '').replace(/\n/g, '').replace(/\t/g, '').trim();

export default cleanString;
