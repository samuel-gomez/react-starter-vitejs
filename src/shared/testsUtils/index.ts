export * from './customRender';
export { default as emptyFunction } from '../helpers/emptyFunction';
export { default as MOCK_API_URL } from './constants';
export { default as renderWithContainer } from './renderWithContainer';
