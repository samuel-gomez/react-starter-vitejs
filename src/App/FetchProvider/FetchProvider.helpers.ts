import { useOidcAccessToken } from '@axa-fr/react-oidc';
import { type QueryKey } from '@tanstack/react-query';
import fetch from 'cross-fetch';
import { API_URL, STATUS_API, STATUS_HTTP_MESSAGES } from 'shared/constants';
import { isEmptyOrNull, mergeObj } from 'shared/helpers';
import { useOidcAccessTokenMock } from './constants';

type TsetResponseError = {
  response: {
    anomaly?: {
      label?: string;
      detail?: string;
    };
    status: number;
  };
};

export const setResponseError = ({ response }: TsetResponseError) => {
  const { anomaly, status } = response;
  switch (true) {
    case `${status}`.startsWith(`${STATUS_API.WARNING}`):
      return {
        ...anomaly,
        label: isEmptyOrNull(anomaly?.label) ? STATUS_HTTP_MESSAGES[status] : anomaly?.label,
        detail: anomaly?.detail ?? '',
        type: 'danger',
        iconName: 'alert',
      };
    case `${status}`.startsWith(`${STATUS_API.ERROR}`):
      return {
        ...anomaly,
        label: isEmptyOrNull(anomaly?.label) ? STATUS_HTTP_MESSAGES[status] : anomaly?.label,
        detail: anomaly?.detail ?? '',
      };
    default:
      return {};
  }
};

export const computeDataError = async (response: Response, setResponseErrorFn = setResponseError) => {
  try {
    const data = await response.json();
    return setResponseErrorFn({ response: { ...data, status: response.status } });
  } catch (error) {
    return setResponseErrorFn({ response: { anomaly: { label: STATUS_HTTP_MESSAGES[response.status] }, status: response.status } });
  }
};

type TConfig = Record<string, boolean>;

export const buildResponse = async (response: Response, config: TConfig, computeDataErrorFn = computeDataError) => {
  const { status } = response;

  switch (true) {
    case `${status}`.startsWith(`${STATUS_API.ERROR}`):
    case `${status}`.startsWith(`${STATUS_API.WARNING}`): {
      throw await computeDataErrorFn(response);
    }
    case `${status}` === `${STATUS_API.SUCCESS}`:
      if (config.blob) {
        return response.blob();
      }
      if (config.text) {
        return response.text();
      }
      return {
        ...(await response.json()),
        statusHttp: status,
      };
    default:
      return {
        statusHttp: status,
      };
  }
};

export const manageConfig = (apiName: string, fetchAuthConfig: { headers?: object }) => {
  const { headers, ...restFetchAuthConfig } = fetchAuthConfig;
  return apiName === API_URL.BASE && headers ? { headers, ...restFetchAuthConfig } : restFetchAuthConfig;
};

export const catchErrorServer = (error: { message: string }) =>
  Response.json(
    {
      anomaly: {
        label: error.message,
      },
    },
    {
      status: 500,
      statusText: error.message,
    },
  );

export type TsetFetchCustom = {
  apiUrl: Record<string, string>;
  fetchAuthConfig: object;
  fetchFn?: typeof fetch;
  mergeObjFn?: typeof mergeObj;
  manageConfigFn?: typeof manageConfig;
  catchErrorServerFn?: typeof catchErrorServer;
};

export const setFetchCustom =
  ({
    apiUrl,
    fetchAuthConfig,
    fetchFn = fetch,
    mergeObjFn = mergeObj,
    manageConfigFn = manageConfig,
    catchErrorServerFn = catchErrorServer,
  }: TsetFetchCustom) =>
  async (queryKey: QueryKey) => {
    const [path, customConfig, apiName = API_URL.BASE]: QueryKey = queryKey;
    const url = `${apiUrl[apiName as string]}${path}`;
    const fetchAuthConfigCustom = manageConfigFn(apiName as string, fetchAuthConfig);
    const config = mergeObjFn(fetchAuthConfigCustom, customConfig);

    const response = await fetchFn(url, config).catch(catchErrorServerFn);
    return buildResponse(response, config);
  };

export const getAccessToken = ({
  isEnabled,
  useOidcAccessTokenFn = useOidcAccessToken,
  useOidcAccessTokenMockFn = useOidcAccessTokenMock,
}: {
  isEnabled?: boolean;
  useOidcAccessTokenFn?: typeof useOidcAccessToken;
  useOidcAccessTokenMockFn?: typeof useOidcAccessTokenMock;
}) => (isEnabled ? useOidcAccessTokenFn : useOidcAccessTokenMockFn);
