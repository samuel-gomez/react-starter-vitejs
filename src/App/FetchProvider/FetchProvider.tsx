import type { QueryKey } from '@tanstack/react-query';
import { EnvironmentContext } from 'App/EnvironmentProvider';
import { createContext, useContext, useMemo, type ReactNode } from 'react';
import { mergeObj } from 'shared/helpers';
import { getAccessToken, setFetchCustom } from './FetchProvider.helpers';

export type TFetchCustom = <T>(queryKey: QueryKey) => Promise<T>;

export type FetchContextType = {
  fetchCustom: TFetchCustom;
};

export const FetchContext = createContext<FetchContextType | object>({});
FetchContext.displayName = 'FetchContext';

export type TFetchProvider = {
  mergeObjFn?: typeof mergeObj;
  children: ReactNode;
  getAccessTokenFn?: typeof getAccessToken;
  setFetchCustomFn?: typeof setFetchCustom;
};

const FetchProvider = ({ children, mergeObjFn = mergeObj, setFetchCustomFn = setFetchCustom, getAccessTokenFn = getAccessToken }: TFetchProvider) => {
  const { environment } = useContext(EnvironmentContext);
  const { accessToken } = getAccessTokenFn({ isEnabled: environment?.oidc?.isEnabled })();

  const authConfig = {
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
  };

  const fetchAuthConfig = mergeObjFn(environment?.fetchConfig, authConfig);
  const fetchCustom = setFetchCustomFn({ apiUrl: environment?.apiUrl ?? {}, fetchAuthConfig });
  const value = useMemo(() => ({ fetchCustom }), [fetchCustom]);

  return <FetchContext.Provider value={value}>{children}</FetchContext.Provider>;
};
export default FetchProvider;
