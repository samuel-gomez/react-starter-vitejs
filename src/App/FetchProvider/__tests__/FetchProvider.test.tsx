import { useContext } from 'react';
import { render } from '@testing-library/react';
import EnvironmentProvider from 'App/EnvironmentProvider';
import { MOCK_API_URL } from 'shared/testsUtils';
import FetchProvider, { FetchContext, type TFetchCustom } from '../FetchProvider';

type TBase = {
  fetchCustom?: TFetchCustom;
};

const Base = ({ fetchCustom }: TBase) => <div>{fetchCustom ? 'haveFetchCustom' : 'notHaveFetchCustom'}</div>;

const BaseWithFetch = () => {
  const fetchProps = useContext(FetchContext);
  return <Base {...fetchProps} />;
};

const getAccessTokenFnMock = vi.fn().mockReturnValue(() => ({
  accessToken: 'accessTokenfdsfdsqgvqvsqfs',
}));

describe('FetchProvider', () => {
  it('Should Base have fetchCustom props when render FetchProvider with required props', async () => {
    const useEnvFn = vi.fn().mockReturnValueOnce({
      envState: {
        environment: {
          apiUrl: MOCK_API_URL,
          fetchConfig: {},
        },
      },
    });
    const { getByText } = render(
      <EnvironmentProvider useEnvFn={useEnvFn}>
        <FetchProvider getAccessTokenFn={getAccessTokenFnMock}>
          <BaseWithFetch />
        </FetchProvider>
      </EnvironmentProvider>,
    );

    expect(getByText(/haveFetchCustom/)).toBeInTheDocument();
  });

  it('Should setFetchCustomFn to have been called with apiUrl = {} when environment is null', () => {
    const useEnvFn = vi.fn().mockReturnValueOnce({
      envState: {
        environment: null,
      },
    });

    const mergeObjFn = vi.fn().mockReturnValueOnce({});
    const setFetchCustomFn = vi.fn();
    render(
      <EnvironmentProvider useEnvFn={useEnvFn}>
        <FetchProvider getAccessTokenFn={getAccessTokenFnMock} mergeObjFn={mergeObjFn} setFetchCustomFn={setFetchCustomFn}>
          <BaseWithFetch />
        </FetchProvider>
      </EnvironmentProvider>,
    );

    expect(setFetchCustomFn).toHaveBeenCalledWith({ apiUrl: {}, fetchAuthConfig: {} });
  });
});
