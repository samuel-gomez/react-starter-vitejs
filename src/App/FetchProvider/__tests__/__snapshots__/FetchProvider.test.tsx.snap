// Vitest Snapshot v1

exports[`FetchProvider > Should Base have fetchCustom props when render FetchProvider with required props 1`] = `
<DocumentFragment>
  <div>
    haveFetchCustom
  </div>
</DocumentFragment>
`;
